### PR TITLE
Rework AMG experiments

### PR DESCRIPTION
**Pull request by pearce8**
Updated:

@rfhaque Please re-review and merge.

Types of experiments 
- [x] The current Problem 1 experiment should be named "throughput" (varying the size of the problem, keeping resources the same).  x-axis label: problem size.  Add plots here.
- [x] Add a strong scaling experiment.  x-axis label: number of GPUs.  Add plots here.
- [x] Write a weak scaling experiment.  x-axis label: number of GPUs.  Add plots here.
- [x] Add corresponding experiments for Problem 2 (throughput, strong, weak)

What is working:
- [x] CUDA on lassen
- [x] ROCM on tioga
- [x] OpenMP on ruby: throughput works, need params for strong and weak

### CODE REVIEW

**Comment by Thionazin** [comment]
amg_p2_omp_weak_tot
amg_p2_omp_weak_perc
amg_p2_omp_strong_tot
amg_p2_omp_strong_perc
amg_p1_omp_weak_tot
amg_p1_omp_weak_perc
amg_p1_omp_strong_tot
amg_p1_omp_strong_perc

**Comment by Thionazin** [comment]
amg_p2_rocm_weak_tot
amg_p2_rocm_weak_perc
amg_p2_rocm_strong_tot
amg_p2_rocm_strong_perc
amg_p1_rocm_weak_tot
amg_p1_rocm_weak_perc
amg_p1_rocm_strong_tot
amg_p1_rocm_strong_perc

**Comment by Thionazin** [comment]
amg_p2_cuda_weak_tot
amg_p2_cuda_weak_perc
amg_p2_cuda_strong_tot
amg_p2_cuda_strong_perc
amg_p1_cuda_weak_tot
amg_p1_cuda_weak_perc
amg_p1_cuda_strong_tot
amg_p1_cuda_strong_perc

**Comment by pearce8** [comment]
@rfhaque Please take a look at how this PR gets around the questions, but we need a longer-term plan.

GTL: GTL is GPU Transport Layer, an HPE library required for CUDA-aware MPI.

[ ]  Currently using GTL everywhere.  Is this correct?
[ ]  We are not using CrayMPI on lassen - should we use the GTL or non-GTL version?
[ ]  What version should we use on Pascal (also no CrayMPI)
[ ]  Rocm experiments: do we need GTL?  Make them match the Cuda experiments

**Review by pearce8 — COMMENTED** [review]
@rfhaque the rest looks good to me.  If you approve the way GTL is handled, we can merge this PR.

File: experiments/amg2023/cuda/ramble.yaml [hunk 1]
```diff
@@ -22,32 +22,81 @@ ramble:
...
               variables:
-                env_name: amg2023
+                px: '2'
+                py: '2'
+                pz: '2'
+                nx: ['10', '20', '20']
+                ny: ['10', '10', '20']
+                nz: ['10', '10', '10']
+                env_name: 'amg2023'
                 n_gpus: '8'
```
Comment (rfhaque) [inline experiments/amg2023/cuda/ramble.yaml]:
```suggestion
                n_gpus: '{px}*{py}*{pz}'
```

File: experiments/amg2023/cuda/ramble.yaml [hunk 2]
```diff
@@ -22,32 +22,81 @@ ramble:
...
+                package_manager: spack
+              variables:
+                px: '2'
+                py: '2'
+                pz: '2'
+                nx: ['10', '20', '20']
+                ny: ['10', '10', '20']
+                nz: ['10', '10', '10']
+                env_name: 'amg2023'
+                n_gpus: '8'
```
Comment (rfhaque) [inline experiments/amg2023/cuda/ramble.yaml]:
```suggestion
                n_gpus: '{px}*{py}*{pz}'
```

File: experiments/amg2023/rocm/ramble.yaml [hunk 3]
```diff
@@ -19,46 +19,101 @@ ramble:
   - name: allocation
 
   variables:
-    gtl: ["gtl", "no-gtl"]
+    gtl: "gtl"
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]: @pearce8 We can remove the `gtl` variable completely

File: experiments/amg2023/rocm/ramble.yaml [hunk 4]
```diff
@@ -19,46 +19,101 @@ ramble:
...
               variants:
                 package_manager: spack
               variables:
+                px: '2'
+                py: '2'
+                pz: '2'
+                nx: ['10', '20', '20']
+                ny: ['10', '10', '20']
+                nz: ['10', '10', '10']
                 env_name: 'amg2023-gpu-{gtl}'
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]:
```suggestion
                env_name: 'amg2023-gpu'
```

File: experiments/amg2023/rocm/ramble.yaml [hunk 5]
```diff
@@ -19,46 +19,101 @@ ramble:
...
+              variants:
+                package_manager: spack
+              variables:
+                px: ['2', '2', '4']
+                py: ['2', '2', '2']
+                pz: ['1', '2', '2']
+                nx: ['10', '20', '20']
+                ny: ['10', '10', '20']
+                nz: ['10', '10', '10']
+                env_name: 'amg2023-gpu-{gtl}'
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]:
```suggestion
                env_name: 'amg2023-gpu'
```

File: experiments/amg2023/rocm/ramble.yaml [hunk 6]
```diff
@@ -19,46 +19,101 @@ ramble:
...
+              variants:
+                package_manager: spack
+              variables:
+                px: '2'
+                py: '2'
+                pz: '2'
+                nx: ['10', '20', '20']
+                ny: ['10', '10', '20']
+                nz: ['10', '10', '10']
+                env_name: 'amg2023-gpu-{gtl}'
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]:
```suggestion
                env_name: 'amg2023-gpu'
```

File: experiments/amg2023/rocm/ramble.yaml [hunk 7]
```diff
@@ -19,46 +19,101 @@ ramble:
...
+              variants:
+                package_manager: spack
+              variables:
+                px: ['2', '2', '4']
+                py: ['2', '2', '2']
+                pz: ['1', '2', '2']
+                nx: '10'
+                ny: '10'
+                nz: '10'
+                env_name: 'amg2023-gpu-{gtl}'
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]:
```suggestion
                env_name: 'amg2023-gpu'
```

File: experiments/amg2023/rocm/ramble.yaml [hunk 8]
```diff
@@ -19,46 +19,101 @@ ramble:
...
+              variants:
+                package_manager: spack
+              variables:
+                px: ['2', '2', '4']
+                py: ['2', '2', '2']
+                pz: ['1', '2', '2']
+                nx: ['10', '20', '20']
+                ny: ['10', '10', '20']
+                nz: ['10', '10', '10']
+                env_name: 'amg2023-gpu-{gtl}'
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]:
```suggestion
                env_name: 'amg2023-gpu'
```

File: experiments/amg2023/rocm/ramble.yaml [hunk 9]
```diff
@@ -19,46 +19,101 @@ ramble:
...
+                n_gpus: '{px}*{py}*{pz}'
   software:
     packages:
-      hypre-{gtl}:
+      hypre-gtl:
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]:
```suggestion
      hypre:
```

File: experiments/amg2023/rocm/ramble.yaml [hunk 10]
```diff
@@ -19,46 +19,101 @@ ramble:
...
         pkg_spec: hypre@2.31.0 +mpi+rocm+mixedint~fortran{modifier_spack_variant} amdgpu_target={rocm_arch}
         compiler: compiler-rocm
-      amg2023-gpu-{gtl}:
+      amg2023-gpu-gtl:
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]:
```suggestion
      amg2023-gpu:
```

File: experiments/amg2023/rocm/ramble.yaml [hunk 11]
```diff
@@ -19,46 +19,101 @@ ramble:
...
         pkg_spec: amg2023@develop +mpi+rocm{modifier_spack_variant} amdgpu_target={rocm_arch}
         compiler: compiler-rocm
     environments:
-      amg2023-gpu-{gtl}:
+      amg2023-gpu-gtl:
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]:
```suggestion
      amg2023-gpu:
```

File: experiments/amg2023/rocm/ramble.yaml [hunk 12]
```diff
@@ -19,46 +19,101 @@ ramble:
...
         packages:
         - blas-rocm
         - lapack
-        - mpi-rocm-{gtl}
-        - hypre-{gtl}
-        - amg2023-gpu-{gtl}
+        - mpi-rocm-gtl
+        - hypre-gtl
+        - amg2023-gpu-gtl
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]:
```suggestion
        - mpi-rocm
        - hypre
        - amg2023-gpu
```

File: experiments/amg2023/rocm/ramble.yaml [hunk 13]
```diff
@@ -19,46 +19,101 @@ ramble:
...
                 package_manager: spack
               variables:
+                px: '2'
+                py: '2'
+                pz: '2'
+                nx: ['10', '20', '20']
+                ny: ['10', '10', '20']
+                nz: ['10', '10', '10']
                 env_name: 'amg2023-gpu-{gtl}'
                 n_gpus: '8'
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]:
```suggestion
                n_gpus: '{px}*{py}*{pz}'
```

File: experiments/amg2023/rocm/ramble.yaml [hunk 14]
```diff
@@ -19,46 +19,101 @@ ramble:
...
+                package_manager: spack
+              variables:
+                px: '2'
+                py: '2'
+                pz: '2'
+                nx: ['10', '20', '20']
+                ny: ['10', '10', '20']
+                nz: ['10', '10', '10']
+                env_name: 'amg2023-gpu-{gtl}'
+                n_gpus: '8'
```
Comment (rfhaque) [inline experiments/amg2023/rocm/ramble.yaml]:
```suggestion
                n_gpus: '{px}*{py}*{pz}'
```

File: experiments/amg2023/openmp/ramble.yaml [hunk 15]
```diff
@@ -22,28 +22,95 @@ ramble:
...
               variables:
+                p: 2
+                px: '{p}'
+                py: '{p}'
+                pz: '{p}'
+                nx: ['10', '20', '20']
+                ny: ['10', '10', '20']
+                nz: ['10', '10', '10']
+                n_ranks_per_node: '8'
+                n_nodes: '1'
```
Comment (rfhaque) [inline experiments/amg2023/openmp/ramble.yaml]:
```suggestion
                n_nodes: '{px}*{py}*{pz}/{n_ranks_per_node}'
```
Not sure if this works in ramble

**Comment by rfhaque** [comment]
> @rfhaque Please take a look at how this PR gets around the questions, but we need a longer-term plan.
> 
> GTL: GTL is GPU Transport Layer, an HPE library required for CUDA-aware MPI [ ] Currently using GTL everywhere. Is this correct? [ ]

We are using gtl on tioga. I think we should use whichever is the default MPI version on each platform (and remove all references to gtl from ramble.yaml

> We are not using CrayMPI on lassen - should we use the GTL or non-GTL version? [ ] 

We use spectrum-mpi (non-gtl) which is the default mpi on lassen

> What version should we use on Pascal (also no CrayMPI) [ ] 

We use mvapich2-2.3.7-gcc-11.2.1 on Pascal, that is not gtl enabled

> Rocm experiments: do we need GTL? Make them match the Cuda experiments

We do not need gtl in the rocm experiments. We can use the default mode provided by the underlying MPI implementation on the platform

**Review by rfhaque — APPROVED** [review]

**Pull request merged**